brainzutils/ratelimit.py

```diff
@@ -0,0 +1,261 @@
+# The original version of this code was written by Armin Ronacher:
+#
+# This snippet by Armin Ronacher can be used freely for anything you like. Consider it public domain.
+#
+# http://flask.pocoo.org/snippets/70/
+#
+from __future__ import absolute_import
+import time
+from functools import update_wrapper
+from flask import request, g
+from brainzutils import cache
+
+# g key for the timeout when limits must be refreshed from cache
+ratelimit_refresh = 60 # in seconds
+ratelimit_timeout = "rate_limits_timeout"
+
+# Defaults
+ratelimit_per_token_default = 50
+ratelimit_per_ip_default = 30
+ratelimit_window_default = 10
+
+# keys
+ratelimit_per_token_key = "rate_limit_per_token_limit"
+ratelimit_per_ip_key = "rate_limit_per_ip_limit"
+ratelimit_window_key = "rate_limit_window"
+
+# external functions
+ratelimit_user_validation = None
+
+'''
+    HOW TO USE THIS MODULE:
+
+    This module defines a set of function that allows your to add ratelimiting to your
+    flask app. There are three values to know and set:
+
+       per_token_limit - The number of requests that are allowed for a caller who is 
+            setting an 
+               Authorization: Token <auth token>
+            header. This limit can be different than the limit for rate limiting on an IP basis.
+
+       per_ip_limit - The number of requests that are allowed for a caller who is not
+            providing an Authorization header and is rate limited on their IP address.
+
+       ratelimit_window - The window, in number of seconds, how long long the limits
+            above are applied for. 
+
+    To add ratelimit capabilities to your flask app, follow these steps:
+
+    1. During app creation add these lines:
+
+          from brainzutils.ratelimit import ratelimit, inject_x_rate_headers
+
+          @app.after_request
+          def after_request_callbacks(response):
+              return inject_x_rate_headers(response)
+
+    2. Then apply the ratelimit() decorator to any function that should be rate limited:
+
+         @app.route('/')
+         @ratelimit()
```

Comment:
small thing - if this decorator takes no parameters then it'd be nice to be able to call it as `@ratelimit` without parenthesis: https://stackoverflow.com/questions/35572663/using-python-decorator-with-or-without-parentheses